Add Versioned Component Configuration Files doc (#6758)
kubernetes/community: archive/wg-component-standard/component-config/README.md

> This copy of [Versioned Component Configuration Files](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c/)

Reviewer: Consider using the self-link

```suggestion
This copy of [Versioned Component Configuration Files](https://goo.gl/GM8KyH)
```

or

```suggestion
This copy of [Versioned Component Configuration Files][vccf-proposal]
...

[vccf-proposal]: https://goo.gl/GM8KyH
```

if you want to shorten up this line.

Author: Seems like that short URL still works, but there's no way to manage it. Let's keep the full link since it's probably more durable.

was automatically converted from Google Docs to Markdown
so that it could be included in the community archive.
Several interesting comment threads remain unexported but
available in the Google Doc. You can gain access to the

> doc by joining the `dev@kubernetes.io` mailing list (
> previously `kubernetes-dev@googlegroups.com`). Please

Reviewer: ```suggestion
doc by joining the `dev@kubernetes.io` mailing list
(previously `kubernetes-dev@googlegroups.com`). Please
```

do not request access via the Google Docs UI, as this
spams the owners with access requests. Thank you :).

# Versioned Component Configuration Files

## How Kubelet learned to stop using flags and love versioned config. How your component can, too.

**Shared publicly!**

**Author:** mtaufen@google.com

> **Last Updated: **March 28, 2018

Reviewer: ```suggestion
**Last Updated:** March 28, 2018
```

to maintain emphasis

This document restates the motivation and records lessons from OSS work over the past year:

*   a brief review of why we want versioned config files for all core cluster components
*   the _ideal_ state of a component's configuration API
*   how to migrate an existing component to versioned configuration files (Kubelet example)
*   remaining work

# Why versioned config files?

The short answer is that flags are nonstandard interfaces with weak stability guarantees. They are confusing and hard to deploy, and this is the opposite of what Kubernetes should be.

Command-line flags present a number of problems:

*   Flags are a public API, but are not versioned and cannot be versioned separately from the binary: 
    *   For _core_ components the binary version is coupled to the Kubernetes release version. We use semantic versioning for our binaries, but can't bump a major version unless Kubernetes does:
        *   We _shouldn't_ ever **fix bad defaults** for existing flag names without bumping the major version of a binary. In reality, we get around this by giving advance warning and technically breaking semantic versioning of the binary.

>         *   We _shouldn't_ ever **remove a flag **without bumping the major version of a binary. In reality we use a [flag deprecation policy](https://kubernetes.io/docs/reference/deprecation-policy/#deprecating-a-flag-or-cli) that allows us to technically break semantic versioning of the binary as long as we give advance warning.

Reviewer: ```suggestion
        *   We _shouldn't_ ever **remove a flag** without bumping the major version of a binary. In reality we use a [flag deprecation policy](https://kubernetes.io/docs/reference/deprecation-policy/#deprecating-a-flag-or-cli) that allows us to technically break semantic versioning of the binary as long as we give advance warning.
```

to maintain emphasis

*   Enable dynamic configuration deployment mechanisms.

Conveniently, Kubernetes has similar goals:

*   Standardize the configuration approach for cluster infrastructure.
*   Enable dynamic deployment mechanisms.

Kubernetes had already paved the way: It has what we need to version our configuration interfaces, decouple configuration changes from binary changes, represent configuration in a structured format, and deploy configuration in a dynamic environment.

*   Versioning was accomplished via the API machinery's group/version mechanism.
*   Adhering to the same API guarantees as the core Kubernetes APIs provides a stable configuration surface, and allows us to decouple the configuration interface from releases that support the same API version.
*   Kubernetes API objects are defined as Go structs, which means we don't have to parse strings to deal with structured config.
*   Kubernetes has deployment mechanisms (ConfigMap volume source) that work well for pushing new versions of configuration files into production.
*   There is no requirement to restart a process when you change a file, unlike flags.

> **All core Kubernetes components should eventually consume their configuration via versioned configuration files, instead of command-line flags. **

Reviewer: ```suggestion
**All core Kubernetes components should eventually consume their configuration via versioned configuration files, instead of command-line flags.**
```

to maintain emphasis

# tl;dr: What should a component do?

**This is the _ideal_ command-line API for every core cluster component:**

> ```

Reviewer: ```suggestion
```console
```

*   parsing _only_ the flags in this flag set

You can find the example of how the Kubelet took back control of its flag set in [Explicit kubelet flags](https://github.com/kubernetes/kubernetes/pull/57613) (see also the follow-up PR, [#58095](https://github.com/kubernetes/kubernetes/pull/58095)).

Many components indiscriminately add the global flag set to their primary flag set via `pflag.CommandLine.AddGoFlagSet(flag.CommandLine)`. Further, most components delegate their flag parsing, help text generation, etc. to [Cobra](https://github.com/spf13/cobra). Cobra implicitly adds flag.CommandLine in several cases, which unfortunately gets in the way of explicit control over the flag API.

In order for Cobra to parse flags for you, it has to be made aware of your flagset. This is achieved by registering flags to the command's flag set. Cobra will implicitly merge the global command lines with this flag set when it parses flags. Consider the following Go program, which creates a local, isolated flag set (as recommended above). The program also registers a global flag, which we _hope_ won't be parsed, because it is not explicitly registered with our local flag set. 

> ```

Reviewer: ```suggestion
```go
```

package main

import (
	"fmt"
	"os"

	"github.com/spf13/cobra"
	"github.com/spf13/pflag"
)

const use = "testcmd"

var (
	globalFlagTarget string
	localFlagTarget  string
)

func init() {
	pflag.StringVar(&globalFlagTarget, "global-flag", globalFlagTarget, "globally-registered flag")
}

func NewLocalFlagSet() *pflag.FlagSet {
	fs := pflag.NewFlagSet(use, pflag.ContinueOnError)
	fs.StringVar(&localFlagTarget, "local-flag", localFlagTarget, "locally-registered flag")
	return fs
}

func NewTestCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use: use,
		Run: func(cmd *cobra.Command, args []string) {
			fmt.Printf("globalFlagTarget: %q\n", globalFlagTarget)
			fmt.Printf("localFlagTarget: %q\n", localFlagTarget)
		},
	}
	cmd.Flags().AddFlagSet(NewLocalFlagSet())
	return cmd
}

func main() {
	cmd := NewTestCmd()
	if err := cmd.Execute(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}
```

If we run the program, we see that this is not the case. The global flag is parsed by Cobra:

> ```

Reviewer: ```suggestion
```console
```

$ testcmd --global-flag hello

globalFlagTarget: "hello"
localFlagTarget: ""
```

You can circumvent this by disabling Cobra's flag parsing. This, unfortunately, requires that you both parse flags and implement --help short-circuiting on your own. The next example attempts to extend the above to do so.

> ```

Reviewer: ```suggestion
```go
```

	cmd := &cobra.Command{
		Use:                use,
		DisableFlagParsing: true,
		Run: func(cmd *cobra.Command, args []string) {
			// parse our local flag set
			if err := localFlagSet.Parse(args); err != nil {
				cmd.Usage()
				fatal(err)
			}
			// --help
			help, err := localFlagSet.GetBool("help")
			if err != nil {
				fatal(fmt.Errorf(`"help" flag is non-bool, programmer error, please correct`))
			}
			if help {
				cmd.Help()
				return
			}
			// print the flag values
			fmt.Printf("globalFlagTarget: %q\n", globalFlagTarget)
			fmt.Printf("localFlagTarget: %q\n", localFlagTarget)
		},
	}
	localFlagSet.BoolP("help", "h", false, fmt.Sprintf("help for %s", cmd.Name()))
	// Cobra still needs to be aware of our flag set to generate usage and help text
	cmd.Flags().AddFlagSet(localFlagSet)
	return cmd
}

func fatal(err error) {
	fmt.Fprintf(os.Stderr, "error: %v\n", err)
	os.Exit(1)
}

func main() {
	cmd := NewTestCmd()
	if err := cmd.Execute(); err != nil {
		fatal(err)
	}
}
```

When we run the program and attempt to set the global flag, we see that the global flag is now rejected, but we also see that it is still included in the usage text!

> ```

Reviewer: ```suggestion
```console
```

$ testcmd --global-flag hello
Usage:
  testcmd [flags]

Flags:
      --global-flag string   globally-registered flag
  -h, --help                 help for testcmd
      --local-flag string    locally-registered flag
error: unknown flag: --global-flag
```

The same thing happens when we pass --help:

> ```

Reviewer: ```suggestion
```console
```

$ testcmd --help
Usage:
  testcmd [flags]

Flags:
      --global-flag string   globally-registered flag
  -h, --help                 help for testcmd
      --local-flag string    locally-registered flag
```

This is because Cobra also uses the global flags when generating usage and help text. This can be circumvented by doing-it-yourself, again.

> ```

Reviewer: ```suggestion
```go
```

				cmd.Usage()
				fatal(err)
			}
			// --help
			help, err := localFlagSet.GetBool("help")
			if err != nil {
				fatal(fmt.Errorf(`"help" flag is non-bool, programmer error, please correct`))
			}
			if help {
				cmd.Help()
				return
			}
			// print the flag values
			fmt.Printf("globalFlagTarget: %q\n", globalFlagTarget)
			fmt.Printf("localFlagTarget: %q\n", localFlagTarget)
		},
	}
	localFlagSet.BoolP("help", "h", false, fmt.Sprintf("help for %s", cmd.Name()))

	// ugly, but necessary, because Cobra's default UsageFunc and HelpFunc pollute the flagset with global flags
	const usageFmt = "Usage:\n  %s\n\nFlags:\n%s"
	cmd.SetUsageFunc(func(cmd *cobra.Command) error {
		fmt.Fprintf(cmd.OutOrStderr(), usageFmt, cmd.UseLine(), localFlagSet.FlagUsagesWrapped(2))
		return nil
	})
	cmd.SetHelpFunc(func(cmd *cobra.Command, args []string) {
		fmt.Fprintf(cmd.OutOrStdout(), usageFmt, cmd.UseLine(), localFlagSet.FlagUsagesWrapped(2))
	})

	return cmd
}

func fatal(err error) {
	fmt.Fprintf(os.Stderr, "error: %v\n", err)
	os.Exit(1)
}

func main() {
	cmd := NewTestCmd()
	if err := cmd.Execute(); err != nil {
		fatal(err)
	}
}
```

And now things are more as we expect.

> ```

Reviewer: ```suggestion
```console
```
